kafka-client/src/main/java/dev/responsive/kafka/api/WindowStoreOptions.java

```diff
@@ -0,0 +1,134 @@
+package dev.responsive.kafka.api;
+
+import static dev.responsive.kafka.store.StoreUtils.durationToMillis;
+
+import java.time.Duration;
+import java.util.Objects;
+import org.apache.kafka.common.serialization.Serde;
+import org.apache.kafka.streams.processor.StateStore;
+import org.apache.kafka.streams.state.WindowStore;
+
+public class WindowStoreOptions<K, V> {
```

Comment:
This might be better as a "face to face" conversation because we might both feel strongly here, but I have to (respectively 😋) disagree about avoiding convenience in favor of replicating the existing `Stores` API. I don't think I can stress enough just how incredibly awkward, frustrating, and difficult to understand the `Stores` factory and surrounding context really is. 

First of all, there are a million different methods on `Stores` and more are added all the time -- each new store type means (at the least) one supplier and a corresponding builder, and the docs aren't really clear enough about when to use which. There's also a lot of gotchas, for example there are timestamped and "normal" variants of each supplier and each builder, and you have to think carefully about which suppliers can and can't be passed into which builders. 

But (besides stream-stream joins -- see other comment) there's actually no top-level usage for store suppliers -- they are solely for configuring either a StoreBuilder (for PAPI) or Materialized (for DSL) which isn't even part of the `Stores` API.  Why not just offer a store factory that provides the two actual store provider-types, and save users the headache of figuring out which store supplier to use and then what to actually do with it?

Regarding the Options classes, the benefit here is that users don't have to figure out the value of all those parameters and can ignore everything they don't want to override and let us figure out the correct values. This can honestly be infuriating when trying to customize a window store, because for some operators the parameters are flexible  while others expect a certain value and only throw an exception at runtime if you gave the wrong thing. And it's not obvious or even easy to figure out what the right value is -- especially stream-stream joins. I had to work it out through trial-and-error because the error message doesn't even tell you (I actually did [contribute a PR](https://github.com/apache/kafka/pull/13682/) to improve this but haven't had time to finish fixing up the tests )

So it's hard to figure out which store suppliers to use, how to hook them into your actual Topology depending on DSL vs PAPI, and hard to figure out the value of parameters that you don't need to care about. This would already be bad even if it was "familiar", but sure, some users are already doing this and could substitute our stores for whatever they had before. But a lot of users don't customize their stores at all right now! So migrating to Responsive will be their first experience with this mess, and even though we just stuck to Kafka Streams's API they're going to view it as a problem with _our_ API, because we're the ones forcing them into this. This is the ultimate reason I want to present a cleaner API (and why I wrote this manifesto 😅 )

Reply:
Also, just to put the whole argument out there, it's nice to wrap configs in one options class for compatibility and our ability to add new optional features without rapid expansion of the store factory API. John actually proposed a new grammar for Streams based on this idea, but it never went anywhere (partly pushback from Matthias, partly it was going to be a huge project and disruption for users to migrate)

On that note...with all that said, I of course agree wholeheartedly with your opening line:

> this seems like something that should just be directly in streams, right?

But getting a good clean API in Streams is going to take forever. It's almost 100% touching on the public API which is the nittiest part of every KIP discussion so that will drag out, it'll involve many large PRs that will add nontrivial review load and won't be prioritized since it's  purely cosmetic within AK Streams, and the surface area of the whole thing is huge so there will be a lot of compatibility concerns about whether this is necessary (a fair question imo). 

I definitely DO want to clean up the Streams API, but imo this is probably one of the most difficult things to contribute back. Whereas we can benefit from it right away, like next week. Maybe this is overly pessimistic based on how the grammar proposal from John went, but you can probably imagine how this would go. So I just don't think it makes sense for us pass on improving the user experience until it gets into Streams

Reply:
I apologize if that came off as too aggressive -- I'm not frustrated with you, I'm frustrated with Kafka Streams's stores 😂

Reply:
Definitely let's chat face-to-face on this one. Clearly there's a lot of history here I'm not aware of - I 100% agree that this proposal is better than what's there in streams, that was not in question (at least on face value, admittedly I'm ignorant about the nuance of what's in streams today). It's a shame we can't just contribute this to Streams...

I think what it boils down to is this comment:

> This would already be bad even if it was "familiar", but sure, some users are already doing this and could substitute our stores for whatever they had before. But a lot of users don't customize their stores at all right now! 

If we're confident many people don't customize their stores at all, then that's a good argument to provide a better API for us.

I'm thinking specifically of our first use case, where they do use the stores and they'll need to change their code structure a little to use the new API - not the end of the world, but every change increases the migration guide length (both literal word count and time to migrate). To me, that is a critical metric.

The other thing to consider is that we don't have docs or examples published anywhere - there's a lot of public material out there about Kafka Streams, which we'd have to replicate somehow.

Reply:
my 2c having just waded into this and still catching up on the state store apis - I'm very torn here. I have managed not to really look at the state store API in all my years at Confluent, and after trying to take a pass over it just now my head is spinning. What you've proposed here is way cleaner @ableegoldman.

I think at the end of the day both the use case where an existing user doesn't customize state stores and does customize state stores are important for us to solve. I doubt we can just write off either scenario as being sufficiently rare.

What's the cost of keeping around something that mirrors the existing API, while also offering our own better API for users that just want the basic stores? Is it really that much extra overhead? It might be worth it to keep the migration experience palatable.

Reply:
> What's the cost of keeping around something that mirrors the existing API, while also offering our own better API for users that just want the basic stores? Is it really that much extra overhead?

I think that should be possible, especially b/c our new API can just delegate the old one, and then we get the best of both worlds. It is more API surface area and more mental overhead but I think optimizing for ease of transition for both of those use cases makes sense.